openshift/console
# POC: Basic cloud shell flow

This PR is a combination of the changes from https://github.com/openshift/console/pull/5332 and branch https://github.com/openshift/console/compare/master...sleshchenko:proxiedTerminal, with some minor cleanup.

### Testing
1. Deploy changes to cluster (e.g. via https://github.com/openshift/console#openshift-with-authentication)
2. Deploy che-workspace-operator to cluster:
    ```bash
    git clone https://github.com/che-incubator/che-workspace-operator.git
    cd che-workspace-operator
    export WEBHOOK_ENABLED=true
    export DEFAULT_ROUTING=basic
    make deploy
    ```
3. Login as a regular user (i.e. not kubeadmin)
4. Use terminal button to start up a cloud shell

    ![Screenshot from 2020-05-11 15-44-42](https://user-images.githubusercontent.com/16168279/81605987-4fbc1c80-93a0-11ea-8809-057f97ed5880.png)

    known issue at the moment: first load of the panel will result in an empty box; currently need to close panel and reopen to get shell.

5. Check that you're logged in to the terminal (`oc whoami`)

### Details
The current implementation uses the following flow:
1. Frontend creates a namespace (if necessary) and initializes a Che workspace
2. The che-workspace-operator creates a deployment for the requested workspace
3. The frontend POSTs the user's token to the backend proxy, which forwards it to the workspace's `/exec/init` endpoint
4. The workspace component creates a kubeconfig file for the passed token
5. The front end does a regular `exec` into the workspace container

### Security concerns
We're adding a proxy to the backend that is submitting a user's auth token to an URL defined in a status field. As a result it's necessary to carefully consider cases to avoid sending the auth token to an unintended recipient. The current security model on the workspace operator side is:
- Upon creation, workspaces get a creator annotation (`org.eclipse.che.workspace/creator`) that is set to the user's UID. This is protected by a webhook, and the annotation is read on the backend to ensure the token is being sent to a workspace that was created by the current user.
- Exec access to any containers in a workspace is restricted to the creator, via webhook

Two concerns that have been brought up are:
- If the workspace operator is not deployed, someone could create a workspace and intercept a user's token
  - URL that backend uses for proxy is obtained from workspace/status, which normal users cannot modify. As a result, without the operator running to set a URL, this shouldn't be possible as the target URL would be blank.
- If the workspace operator was once deployed but is removed, workspaces are no longer managed and could potentially be modified to steal a user's token.
  - Currently the only validation on the front-end is that a frontend-specific annotation is applied: `console.openshift.io/cloudshell-user: developer`; this would need to be updated to avoid this issue
  - One approach is to always create a new workspace rather than fetching an existing workspace, though the time this takes would likely make the feature unviable

I don't have a definitive answer to these concerns, at the moment.

cc: @sleshchenko
